File: JAVA/2416. Sum of Prefix Scores of Strings.java
class Solution {
    class TrieNode {
        HashMap<Character, TrieNode> children;
        int count; 
        TrieNode() {
            children = new HashMap<>();
            count = 0;
        }
    }
    private void insert(TrieNode root, String word) {
        TrieNode currentNode = root;
        for (char c : word.toCharArray()) {
            currentNode.children.putIfAbsent(c, new TrieNode());
            currentNode = currentNode.children.get(c);
            currentNode.count++; 
        }
    }
    private int getScore(TrieNode root, String word) {
        TrieNode currentNode = root;
        int score = 0;
        for (char c : word.toCharArray()) {
            currentNode = currentNode.children.get(c);
            score += currentNode.count;
        }
        return score;
    }
    public int[] sumPrefixScores(String[] words) {
        TrieNode root = new TrieNode();
        for (String word : words) {
            insert(root, word);
        }
        int[] result = new int[words.length];
        for (int i = 0; i < words.length; i++) {
            result[i] = getScore(root, words[i]);
        }        
        return result;
    }
}

File: C++/2416. Sum of Prefix Scores of Strings.cpp
class Solution {
public:
    struct trie{
        vector<trie*>child;
        int score;
        trie(){
            child.resize(26);
            score=0;
        }
    };    
    void insert(trie* root, string &str){
        for(int i=0;i<str.length();i++){
            if(root->child[str[i]-97]==NULL){
                root->child[str[i]-97]=new trie();
            }
            root=root->child[str[i]-97];
            root->score++;
        }
    }    
    vector<int>res;
    void findscore(trie* root,string &str){
        trie* curr=root;
        int sc=0;
        for(int i=0;i<str.length();i++){
            curr=curr->child[str[i]-97];
            sc+=(curr->score);
        }
        res.push_back(sc);
    }
    vector<int> sumPrefixScores(vector<string>& words) {
       trie *root=new trie();
       for(auto x:words)insert(root,x);
       for(auto x:words)findscore(root,x);
       return res;
    }
};
